modules/aruco/src/aruco.cpp

```diff
@@ -0,0 +1,1573 @@
+    //    int currId = i;
+    //    Mat currentCandidate = _candidates.getMat(i);
+    //    if (_identifyOneCandidate(dictionary, grey, currentCandidate, currId, params)) {
+    //        validCandidates[i] = 1;
+    //        idsTmp[i] = currId;
+    //    }
+    //}
+
+    // this is the parallel call for the previous commented loop (result is equivalent)
+    parallel_for_(Range(0, ncandidates),
+                  IdentifyCandidatesParallel(&grey, _candidates, _contours, &dictionary, &idsTmp,
+                                             &validCandidates, &params));
+
+    for(int i = 0; i < ncandidates; i++) {
+        if(validCandidates[i] == 1) {
+            accepted.push_back(_candidates.getMat(i));
+            ids.push_back(idsTmp[i]);
+        } else {
+            rejected.push_back(_candidates.getMat(i));
+        }
+    }
+
+    // parse output
+    _accepted.create((int)accepted.size(), 1, CV_32FC2);
+    for(unsigned int i = 0; i < accepted.size(); i++) {
+        _accepted.create(4, 1, CV_32FC2, i, true);
+        Mat m = _accepted.getMat(i);
+        accepted[i].copyTo(m);
+    }
+
+    _ids.create((int)ids.size(), 1, CV_32SC1);
+    for(unsigned int i = 0; i < ids.size(); i++)
+        _ids.getMat().ptr< int >(0)[i] = ids[i];
+
+    if(_rejected.needed()) {
+        _rejected.create((int)rejected.size(), 1, CV_32FC2);
+        for(unsigned int i = 0; i < rejected.size(); i++) {
+            _rejected.create(4, 1, CV_32FC2, i, true);
+            Mat m = _rejected.getMat(i);
+            rejected[i].copyTo(m);
+        }
+    }
+}
+
+
+/**
+  * @brief Final filter of markers after its identification
+  */
+static void _filterDetectedMarkers(InputArrayOfArrays _inCorners, InputArray _inIds,
+                                   OutputArrayOfArrays _outCorners, OutputArray _outIds) {
+
+    CV_Assert(_inCorners.total() == _inIds.total());
+    if(_inCorners.total() == 0) return;
+
+    // mark markers that will be removed
+    vector< bool > toRemove(_inCorners.total(), false);
+    bool atLeastOneRemove = false;
+
+    // remove repeated markers with same id, if one contains the other (doble border bug)
+    for(unsigned int i = 0; i < _inCorners.total() - 1; i++) {
+        for(unsigned int j = i + 1; j < _inCorners.total(); j++) {
+            if(_inIds.getMat().ptr< int >(0)[i] != _inIds.getMat().ptr< int >(0)[j]) continue;
+
+            // check if first marker is inside second
+            bool inside = true;
+            for(unsigned int p = 0; p < 4; p++) {
+                Point2f point = _inCorners.getMat(j).ptr< Point2f >(0)[p];
+                if(pointPolygonTest(_inCorners.getMat(i), point, false) < 0) {
+                    inside = false;
+                    break;
+                }
+            }
+            if(inside) {
+                toRemove[j] = true;
+                atLeastOneRemove = true;
+                continue;
+            }
+
+            // check the second marker
```

Comment:
you need not check both actually I think. Depends on the speed of pointPolygonTest. I would assume it is more expensive than calculating pixel distances. You could check the diagonal distances perhaps to see which is higher and just check once. Just a thought.